packages/interaction-tracking/src/InteractionTracking.js

```diff
@@ -0,0 +1,341 @@
+        ((subscriberRef: any): SubscriberRef).current === subscriber,
+      'Only one interactions subscriber may be registered at a time.',
+    );
+    ((subscriberRef: any): SubscriberRef).current = subscriber;
+  }
+}
+
+export function track(
+  name: string,
+  callback: Function,
+  threadID: number = DEFAULT_THREAD_ID,
+): any {
+  if (!enableInteractionTracking) {
+    return callback();
+  }
+
+  const interaction: Interaction = {
+    id: interactionIDCounter++,
+    name,
+    timestamp: now(),
+  };
+
+  const prevInteractions = ((interactionsRef: any): InteractionsRef).current;
+
+  // Tracked interactions should stack/accumulate.
+  // To do that, clone the current interactions.
+  // The previous set will be restored upon completion.
+  const interactions = new Set(prevInteractions);
+  interactions.add(interaction);
+  ((interactionsRef: any): InteractionsRef).current = interactions;
+
+  if (enableInteractionTrackingObserver) {
+    // Update before calling callback in case it schedules follow-up work.
+    interaction.__count = 1;
+
+    let caughtError;
+    let didCatch = false;
+    let returnValue;
+    const subscriber = ((subscriberRef: any): SubscriberRef).current;
+
+    if (subscriber !== null) {
+      try {
+        subscriber.onInteractionTracked(interaction);
+        subscriber.onWorkStarted(interactions, threadID);
+      } catch (error) {
+        didCatch = true;
+        caughtError = caughtError || error;
+      }
+    }
+
+    try {
+      returnValue = callback();
+    } catch (error) {
+      didCatch = true;
+      caughtError = caughtError || error;
+    }
+
+    ((interactionsRef: any): InteractionsRef).current = prevInteractions;
+
+    if (subscriber !== null) {
+      try {
+        subscriber.onWorkStopped(interactions, threadID);
+      } catch (error) {
+        didCatch = true;
+        caughtError = caughtError || error;
+      }
+    }
+
+    interaction.__count = ((interaction.__count: any): number) - 1;
+
+    // If no async work was scheduled for this interaction,
+    // Notify subscribers that it's completed.
+    if (subscriber !== null && ((interaction.__count: any): number) === 0) {
+      try {
+        subscriber.onInteractionScheduledWorkCompleted(interaction);
+      } catch (error) {
+        didCatch = true;
+        caughtError = caughtError || error;
+      }
+    }
+
+    if (didCatch) {
+      throw caughtError;
+    } else {
+      return returnValue;
+    }
+  } else {
+    try {
+      return callback();
+    } finally {
+      ((interactionsRef: any): InteractionsRef).current = prevInteractions;
+    }
+  }
+}
+
+export function unsubscribe(subscriber: Subscriber): void {
+  if (enableInteractionTracking && enableInteractionTrackingObserver) {
+    ((subscriberRef: any): SubscriberRef).current = null;
+  }
+}
+
+export function wrap(
+  callback: Function,
+  threadID: number = DEFAULT_THREAD_ID,
+): Function {
```

Comment:
I don't think I understand why `wrap` is so complicated. I was expecting it to be essentially this:

```js
function wrap(callback) {
  retain();
  const continuedInteractions = interactions.current;
  return (...args) => {
    track(continuedInteractions, () => fn(...args));
    release();
  }
}
```

Reply:
I suppose avoiding the extra function call is worthwhile, but I believe semantically they should be identical. See other comment.

Reply:
I don't think wrap is that complicated. This approach avoids an extra function call and additional wrapper function.

Reply:
Because that's kind of fundamental to how `track` and `wrap` work? I'm not sure I understand the question.

Reply:
I suppose I was expecting the semantics to be that it restores and stacks.

Reply:
`wrap` exists so that you can preserve the interactions at a specific time to track async work against them later. If tracking later modified the set of interactions (by accumulating) that could be confusing.

Probably this doesn't matter in many cases. For something like React that batches work– I think it might.

I don't think this is a problem. 😄

Reply:
It doesn’t modify the current set of interactions because it will reset them back to the previous set after it exits.

Reply:
No, I meant that it would modify the _original_ set. So if I wrapped [foo,bar] I would expect my async work to be attributed to [foo,bar]– not [foo,bar,...whatever-else-was-also-active]

Reply:
I think @acdlite's misunderstanding is a good opportunity to rename this function to something that makes it clearer. I could see how if I see `track` and `wrap` I would think that `wrap` is just a convenience for `wrapTracked` since it is highlighting the wrapping mechanism and nothing else.

In reality the purpose of wrap isn't to wrap a function. The primary purpose is to continue where you left off.

`wrapContinuation`? Too long. This is can be used a lot if you don't have auto-wrapping.

Reply:
I suppose there's precedence for calling it `wrap` in Zones.

Reply:
I like the brevity of "track" and "wrap" but I'm not bullish on them if others feel there are more meaningful names.